### DISCUSSION TITLE
Cholesky decomposition error for boundary restricted AuxKernel

### DISCUSSION THREAD

**Discussion by aprilnovak** [idaholab/moose#16562] [Q&A General] [answered]
Hi all,

I'd like to apply the DiffusionFluxAux auxkernel to a MONOMIAL auxvariable of FIRST order on a boundary. However, when I try this, I get the following error:

```
Time Step 1, time = 1, dt = 1
Error! Can only use Cholesky decomposition with symmetric positive definite matrices.
Stack frames: 26
0: libMesh::print_trace(std::ostream&)
1: libMesh::MacroFunctions::report_error(char const*, int, char const*, char const*)
2: libMesh::DenseMatrix<double>::_cholesky_decompose()
3: void libMesh::DenseMatrix<double>::cholesky_solve<double>(libMesh::DenseVector<double> const&, libMesh::DenseVector<double>&)
4: AuxKernelTempl<double>::compute()
5: ComputeElemAuxBcsThread<AuxKernelTempl<double> >::operator()(libMesh::StoredRange<MooseMesh::const_bnd_elem_iterator, BndElement const*> const&)
6: void libMesh::Threads::parallel_reduce<libMesh::StoredRange<MooseMesh::const_bnd_elem_iterator, BndElement const*>, ComputeElemAuxBcsThread<AuxKernelTempl<double> > >(libMesh::StoredRange<MooseMesh::const_bnd_elem_iterator, BndElement const*> const&, ComputeElemAuxBcsThread<AuxKernelTempl<double> >&)
7: void AuxiliarySystem::computeElementalVarsHelper<AuxKernelTempl<double> >(MooseObjectWarehouse<AuxKernelTempl<double> > const&, std::vector<std::vector<MooseVariableFieldBase*, std::allocator<MooseVariableFieldBase*> >, std::allocator<std::vector<MooseVariableFieldBase*, std::allocator<MooseVariableFieldBase*> > > > const&, unsigned int)
8: AuxiliarySystem::computeElementalVars(MooseEnumItem)
9: AuxiliarySystem::compute(MooseEnumItem)
10: FEProblemBase::computeResidualTags(std::set<unsigned int, std::less<unsigned int>, std::allocator<unsigned int> > const&)
11: FEProblemBase::computeResidualInternal(libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&, std::set<unsigned int, std::less<unsigned int>, std::allocator<unsigned int> > const&)
12: FEProblemBase::computeResidualSys(libMesh::NonlinearImplicitSystem&, libMesh::NumericVector<double> const&, libMesh::NumericVector<double>&)
13: NonlinearSystem::solve()
14: FEProblemBase::solve()
15: FEProblemSolve::solve()
16: PicardSolve::solveStep(double, double&, double, double&, bool, std::set<unsigned int, std::less<unsigned int>, std::allocator<unsigned int> > const&)
17: PicardSolve::solve()
18: TimeStepper::step()
19: Transient::takeStep(double)
20: Transient::execute()
21: MooseApp::executeExecutioner()
22: MooseApp::run()
23: ../../../../cardinal-opt(+0x5eaf) [0x56135c9a5eaf]
24: __libc_start_main
25: ../../../../cardinal-opt(+0x60ae) [0x56135c9a60ae]
[0] ./include/libmesh/dense_matrix_impl.h, line 1001, compiled Dec 18 2020 at 12:35:12
application called MPI_Abort(MPI_COMM_WORLD, 1) - process 0
[unset]: write_line error; fd=-1 buf=:cmd=abort exitcode=1
```

After playing around a bit, I've found two different settings that affect whether or not I see this error:

1. If I change the order to `order = CONSTANT`, but keep the boundary restriction (`boundary = '1'`), I don't see the error.
2. If I remove the boundary restriction (remove `boundary='1'`), I don't see the error for _either_ `order = CONSTANT` or `order = FIRST`.

I'm having a hard time trying to reason out whether if I just shouldn't be trying to apply this AuxKernel in this manner, or if there's a bug here. I would have thought that _any_ order elemental auxvariable might complain about a boundary-restricted auxkernel, but `order = CONSTANT` is fine. If I'm doing something erroneous here, I think a more narrowly-focused error message would be helpful.

I've attached a minimal example that shows this error. The error disappears if you (A) change the `avg_flux` order to `CONSTANT`, or (B) remove `boundary='1'` for the `DiffusionFluxAux` kernel.

Thanks,

-April

**Comment by aprilnovak** [comment]
[example.txt](https://github.com/idaholab/moose/files/5726735/example.txt)

**Reply by fdkong** [reply]
This simply means the matrix is not symmetric anymore when the `DiffusionFluxAux` kernel is restricted on boundary.

**Reply by aprilnovak** [reply]
Gotcha. Is that error supposed to kill the program though? I would have thought that there's not really any solves going on with the auxvariables - just compute them from other variables.

**Reply by fdkong** [reply]
Here is the code:

```
  else /* high-order */
    {
      _local_re.resize(_n_local_dofs);
      _local_re.zero();
      _local_ke.resize(_n_local_dofs, _n_local_dofs);
      _local_ke.zero();

      // assemble the local mass matrix and the load
      for (unsigned int i = 0; i < _test.size(); i++)
        for (_qp = 0; _qp < _qrule->n_points(); _qp++)
        {
          ComputeValueType t = _JxW[_qp] * _coord[_qp] * _test[i][_qp];
          _local_re(i) += t * computeValue();
          for (unsigned int j = 0; j < _test.size(); j++)
            _local_ke(i, j) += t * _test[j][_qp];
        }

      // mass matrix is always SPD
      _local_sol.resize(_n_local_dofs);
      _local_ke.cholesky_solve(_local_re, _local_sol);

      _var.setDofValues(_local_sol);
    }
```

It has nothing to do with PETSc solver for the nonlinear system. It is trying to inverse a small element matrix system to get RHS. If this fails, the computation can not keep going.

**Reply by aprilnovak** [reply]
Would an error like the following always catch when this happens? I guess I am wishing that the error message were more instructive.

```
// in AuxKernel constructor?
if (order > CONSTANT && isParamValid("boundary"))
  mooseError("Auxvariable orders higher than CONSTANT are not supported with boundary-restricted AuxKernels");
```

**Reply by aeslaughter** [reply]
You can make it a paramError("boundary", "AuxVariable orders...") and it will report a line number in the error report.

**Reply by fdkong** [reply]
> Would an error like the following always catch when this happens? I guess I am wishing that the error message were more instructive.

> // in AuxKernel constructor?
> if (order > CONSTANT && isParamValid("boundary"))
>  mooseError("Auxvariable orders higher than CONSTANT are not supported with boundary-restricted AuxKernels");


It is a great idea to add these checks. At the same time, I am still trying to understand why the mass matrix is not symmetric when it is restricted on a boundary.

**Reply by aprilnovak** [reply]
@fdkong, any update on why the mass matrix wouldn't be symmetric on a boundary?

For Cardinal, we're currently restricted to using a CONSTANT MONOMIAL to transfer heat flux into nekRS (because nodal auxvariables don't support gradients, and the Cholesky error is preventing me from using FIRST MONOMIAL). Our BISON mesh for the fuel pins are generally much coarser than the nekRS meshes for the surrounding fluid. Being restricted to a constant heat flux in each BISON element is distorting the heat flux going into nekRS so that we're needing to use really fine BISON meshes (for the solid) just to match the results of a standalone reference nekRS solution (i.e. nekRS solving for both the fluid and solid).

Maybe there's another approach I haven't thought of to retain that spatial resolution of the flux coming from MOOSE?

**Reply by fdkong** [reply]
@aprilnovak If you could send me an example, I will look into this.

**Reply by YaqiWang** [reply]
Do we have any test covering this case in MOOSE? The matrix could be rank deficient because we are only assembling it with the quadrature points on the boundary side. We may need the pseudo inverse for this. But I am surprised we do not see any error before if this is the case.

**Reply by aprilnovak** [reply]
Sure, thanks a bunch! I've attached an example here. This will show you that changing from CONSTANT MONOMIAL to FIRST MONOMIAL throws the error.

After thinking about this more, I realize that I have a second question. It relates to my original question, so I thought I'd ask it here, but if it'd be better as a new discussion let me know! I'll illustrate my question with an example:

## The Problem and Reference Solution
I have a 2-region cylinder heat conduction problem. In the center cylinder, I have a super asymmetrical heat source (to exaggerate the results), while in the outer cylinder there is no heat source. Other than that, there is just heat conduction. To get a reference solution with a _single_ application, I ran two different meshes (I actually made new meshes for each, and didn't just use uniform_refine):

Coarse mesh, with 9128 elements:

![Screen Shot 2021-01-25 at 3 57 15 PM](https://user-images.githubusercontent.com/17039662/105771347-090a5480-5f26-11eb-8b78-073b96054e77.png)

Fine mesh, with 235420 elements:

![Screen Shot 2021-01-25 at 3 55 59 PM](https://user-images.githubusercontent.com/17039662/105771369-11fb2600-5f26-11eb-9e0e-c9fef2e36b93.png)

The reference temperature distribution on the super fine mesh looks like:

![Screen Shot 2021-01-25 at 3 58 47 PM](https://user-images.githubusercontent.com/17039662/105771477-39ea8980-5f26-11eb-8718-613652e7f33f.png)

The temperature on the finer reference mesh is visually identical, so I don't repeat it. On the _interface_ between the two cylinders, the maximum and minimum temperatures for the reference cases are:

Mesh                     | Number of elements | Max temp | Min temp
------------------|---------------------|---------| --
Coarse reference | 9128 | 1959.62 | 1340.72
Fine reference      | 235420 | 1961.22 | 1341.78

My point in showing two reference solutions is to prove that the results I'm trying to replicate are converged. And it's important that the coarse mesh of ~9000 elements is very close to the fine result, so for my intent **I consider the 9000 element mesh to be converged**.

## Solving as Two Regions

So, suppose that I want to match those max/min temperatures for the same physics, but I'm solving each cylinder with a different App. This approximates what I'm doing with the nekRS wrapping - one App solves for the inner cylinder, while a different App solves for the outer cylinder. For instance, here are two meshes - the top one would be solved by the master app (where a `MatchedValueBC` is set on the outer surface), while the bottom one would be solved by the sub app (where a `CoupledVarNeumannBC` is set on the inner surface). Same physics, and with iteration, I should match the reference two-region problem.

App 1 domain and solution:

![Screen Shot 2021-01-25 at 4 10 54 PM](https://user-images.githubusercontent.com/17039662/105772542-f2fd9380-5f27-11eb-99f2-9328005d0c78.png)

App 2 domain and solution:

![Screen Shot 2021-01-25 at 4 11 00 PM](https://user-images.githubusercontent.com/17039662/105772553-f729b100-5f27-11eb-9772-f359d334f540.png)


### The Problem:

The problem is that, _even when using meshes of similar resolution_ as the reference meshes, the fact that I'm limited to a CONSTANT MONOMIAL heat flux going in to the outer cylinder means that I need a super fine mesh just to represent the data transfer. This table shows examples of what I'm talking about: I gradually refine the meshes in the `nek_master.i` and `nek.i` input files (I always match them perfectly at the interface to avoid anything weird with needing to normalize incoming flux values). When the number of elements is similar to the number used in the reference mesh, errors are on the order of 25 K. Even when using meshes with 3x as many elements as the reference mesh, errors are still on the order of 8 K. 


Mesh                     | Number of elements | Max temp ERROR | Min temp ERROR
------------------|---------------------|---------| --
2 coupled apps | 9000 | 23.5 | 6.1
2 coupled apps      | 12000 | 13.0 | 3.3
2 coupled apps | 18000 | 9.2 | 2.3
2 coupled apps | 24000 | 8.0 | 1.6

Here's the finest mesh I'm using:

![Screen Shot 2021-01-25 at 4 25 30 PM](https://user-images.githubusercontent.com/17039662/105773718-f6921a00-5f29-11eb-8687-7626e8791248.png)

This is way finer than the coarse mesh I'm comparing against (i.e. picture 1), so I would expect good agreement.

## The Question

How can I solve two coupled Apps and get similar results as the fully-coupled two-region problem, without needing > 3x finer meshes in the coupled App case?

Is it impossible? From my _original_ question, my motivation was getting around this super fine mesh issue, which is really the root of what I'm interested in addressing.

**Reply by YaqiWang** [reply]
Another thing I need to warn you is that it is not right to let the aux kernel on all boundaries because at the corner element, where the solution is obtained or of which sides of the two attaching boundary sides the solution is obtained, is not clear. Refer to https://github.com/idaholab/moose/issues/5061.

**Reply by aprilnovak** [reply]
Here's the files for my example
[nek_master.txt](https://github.com/idaholab/moose/files/5870065/nek_master.txt)
[nek.txt](https://github.com/idaholab/moose/files/5870066/nek.txt)

[NormalDiffusionFluxAux.zip](https://github.com/idaholab/moose/files/5870070/NormalDiffusionFluxAux.zip)

**Reply by YaqiWang** [reply]
This `changing from CONSTANT MONOMIAL to FIRST MONOMIAL throws the error` makes sense actually because we have two dofs on the side for a linear shape but we have total 3 dofs for the element. Maybe we do not have any test using high order elemental variables.

**Reply by aprilnovak** [reply]
Sideset variables would be nice - if there is a way that I can use FIRST MONOMIAL here, I think that would be the fix for me mesh refinement issue.

**Reply by fdkong** [reply]
@aprilnovak I think I understood  what issue you have right now. I will try your example on my side. Will let you know how it goes

**Reply by hugary1995** [reply]
I just wanted to point out that the error (in the solution) here is largely coming from the Neumann BC in the sub app. There is no simple way around mesh refinement here.

For example, I added the following postprocessors in your sub app:
```
  [flux_integral]
    type = SideFluxIntegral
    variable = nek_temp
    diffusivity = 1
    boundary = '0'
  []
  [monomial_flux_integral]
    type = SideIntegralVariablePostprocessor
    variable = avg_flux
    boundary = '0'
  []
```
and I got `flux_integral = -4.852397e+03` and `monomial_flux_integral = 5.014144e+03`. Notice that ideally they should be equal and opposite sign.

@friedmud and @bwspenc had this discussion on slack a while ago. As pointed out by @friedmud, a similar issue was discussed in [this presentation](https://github.com/idaholab/moose/files/5870430/sc_flux.pdf) by @jwpeterson.

**Reply by fdkong** [reply]
Use the conservative transfer to adjust the transferred variables?

**Reply by aprilnovak** [reply]
The errors aren't related to using conservative transfers - you're right, I did omit them in this example, but in our nekRS wrapping we do make sure to conserve flux by doing some integrations in nekRS (and I've been seeing the same issues for our wrapped App).

For instance, I've attached files here with conservative transfers, and the errors for a 9000 element 2-app case are still > 20 K (provided I did this correctly)

[nek_c.txt](https://github.com/idaholab/moose/files/5870592/nek_c.txt)
[nek_master_c.txt](https://github.com/idaholab/moose/files/5870593/nek_master_c.txt)

**Reply by aprilnovak** [reply]
@hugary1995, I think your example is referring to the fact that the heat flux in my sub-app is only weakly-imposed? Yes, that will definitely cause some differences. 

But I was thinking of this more from the perspective of the spatial loss of information that happens during the formation of the CONSTANT MONOMIAL flux. If we could _not_ lose that information, then my intuition tells me that the error would probably be smaller (it would of course still contain some contribution due to the weak imposition of flux).

**Reply by hugary1995** [reply]
@aprilnovak Yeah that's one aspect of it, but that's not the main issue here. 

By splitting your problem into two subapps, this essentially becomes an optimization problem using alternating directions (this is the terminology in the optimization community). Due to the weak imposition of the Neumann BC (both in the main app when you extract the flux, and in the sub all when you enforce the Neumann BC), there is a mismatch in your objective function (or energy), which you can see from the side set integral of flux. 

In a situation like this, there is no guarantee that the solution you obtain is consistent with your original problem.

**Comment by aeslaughter** [comment]
@fdkong Can you help with this?

**Comment by YaqiWang** [comment, marked as answer]
This change can produce the error on my Mac:
```
diff --git a/test/tests/auxkernels/element_aux_boundary/element_aux_boundary.i b/test/tests/auxkernels/element_aux_boundary/element_aux_boundary.i
index 3a8bbd227a..de6b47fe36 100644
--- a/test/tests/auxkernels/element_aux_boundary/element_aux_boundary.i
+++ b/test/tests/auxkernels/element_aux_boundary/element_aux_boundary.i
@@ -13,7 +13,7 @@
 [AuxVariables]
   [./real_property]
     family = MONOMIAL
-    order = CONSTANT
+    order = SECOND
   [../]
 []
```
Supposedly with `order=FIRST` we should see the same error, but I check the code in libMesh,
```
              libmesh_error_msg_if(A(i,j) <= 0.0,
                                   "Error! Can only use Cholesky decomposition with symmetric positive definite matrices.");
```
possibly it is very sensitive to numerical round-off comparing with 0, thus we did not see the error with the first order. Possibly due to the same reason, you did not see errors in some of your calculations. I checked several moose tests, they are all using constant monomial unfortunately. The fix will be using different lapack function to find a special solution for the singular mass matrix. For example, replace
```
_local_ke.cholesky_solve(_local_re, _local_sol);
```
in AuxKernel.C with
```
_local_ke.svd_solve(_local_re, _local_sol);
```
can temporarily solve the problem. We choose `cholesky_solve` because we think mass matrix is SPD and Cholesky is more efficient.

**Reply by aprilnovak** [reply]
> thus we did not see the error with the first order

I actually did see the error with FIRST order - but it's interesting that you didn't! That's definitely a good insight. In any case, it would probably be a good idea to include some non-CONSTANT monomial tests in MOOSE.

**Reply by YaqiWang** [reply]
Good to hear that. You can try `svd_solve` for now if your task is urgent.

**Reply by YaqiWang** [reply]
Also April, not sure if you noticed my warning in my previous post. I did not check your model, but be sure you do not hit the situation I described in #5061.
